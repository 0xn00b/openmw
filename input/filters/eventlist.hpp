#ifndef MANGLE_INPUT_EVENTLIST_H
#define MANGLE_INPUT_EVENTLIST_H

#include "../event.hpp"
#include <vector>

namespace Mangle
{
  namespace Input
  {
    /** And Event handler that distributes each event to a list of
        other handlers. Supports filtering events by their Type
        parameter.
     */
    struct EventList : Event
    {
      struct Filter
      {
        EventPtr evt;
        int flags;
      };
      std::vector<Filter> list;

      void add(EventPtr e, int flags = EV_ALL)
      {
        Filter f;
        f.evt = e;
        f.flags = flags;
        list.push_back(f);
      }

      virtual void event(Type type, int index, const void *p)
      {
        std::vector<Filter>::iterator it;

        for(it=list.begin(); it!=list.end(); it++)
          {
            if(type & it->flags)
              it->evt->event(type,index,p);
          }
      }
    };
  }
}
#endif
