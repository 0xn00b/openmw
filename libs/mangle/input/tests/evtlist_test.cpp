#include <iostream>
#include "../filters/eventlist.hpp"

using namespace std;
using namespace Mangle::Input;

struct MyEvent : Event
{
  int ii;
  MyEvent(int i) : ii(i) {}

  void event(Event::Type type, int i, const void *p)
  {
    cout << "  #" << ii << " got event: type=" << type << " index=" << i << endl;
  }
};

EventList lst;

int iii=1;
void make(int flags)
{
  lst.add(EventPtr(new MyEvent(iii++)), flags);
}

void send(Event::Type type)
{
  cout << "Sending type " << type << endl;
  lst.event(type,0,NULL);
}

int main()
{
  make(Event::EV_ALL);
  make(Event::EV_KeyDown);
  make(Event::EV_KeyUp | Event::EV_MouseDown);

  send(Event::EV_Unknown);
  send(Event::EV_KeyDown);
  send(Event::EV_KeyUp);
  send(Event::EV_MouseDown);

  cout << "Enough of that\n";
  return 0;
}
