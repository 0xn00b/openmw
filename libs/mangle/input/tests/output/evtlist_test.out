Sending type 1
  #1 got event: type=1 index=0
Sending type 2
  #1 got event: type=2 index=0
  #2 got event: type=2 index=0
Sending type 4
  #1 got event: type=4 index=0
  #3 got event: type=4 index=0
Sending type 16
  #1 got event: type=16 index=0
  #3 got event: type=16 index=0
Enough of that
